conda_libmamba_solver/shards.py

```diff
@@ -0,0 +1,341 @@
+# Copyright (C) 2022 Anaconda, Inc
+# Copyright (C) 2023 conda
+# SPDX-License-Identifier: BSD-3-Clause
+"""
+Models for sharded repodata, and to make monolithic repodata look like sharded
+repodata.
+"""
+
+from __future__ import annotations
+
+import concurrent.futures
+import time
+from collections import defaultdict
+from typing import TYPE_CHECKING, TypedDict
+from urllib.parse import urljoin
+
+import conda.gateways.repodata as repodata
+import msgpack
+import zstandard
+from conda.base.context import context
+from conda.gateways.connection.session import get_session
+from conda.gateways.repodata import (
+    _add_http_value_to_dict,
+    conda_http_errors,
+)
+from conda.models.records import PackageRecord
+from requests import HTTPError
+
+if TYPE_CHECKING:
+    from collections.abc import Iterable
+
+    from conda.core.subdir_data import SubdirData
+    from conda.gateways.repodata import (
+        RepodataCache,
+    )
+    from requests import Response
+
+
+Shard = TypedDict("Shard", {"packages": dict[str, dict], "packages.conda": dict[str, dict]})
+
+
+class RepodataInfo(TypedDict):  # noqa: F811
+    base_url: str  # where packages are stored
+    shards_base_url: str  # where shards are stored
+    subdir: str
+
+
+def maybe_unpack_record(record):
+    """
+    Convert bytes checksums to hex; leave unchanged if already str.
+    """
+    for hash_type in "sha256", "md5":
+        if hash_value := record.get(hash_type):
+            if isinstance(hash_value, bytes):
+                record[hash_type] = hash_value.hex()
+    return record
+
+
+def shard_mentioned_packages(shard: Shard):
+    """
+    Return all dependencies mentioned in a shard, including the shard's own
+    package name.
+
+    Includes virtual packages.
+    """
+    # XXX filter by package name for the possibility of a shard with multiple
+    # (small) packages
+    mentioned = set()
+    for package in (*shard["packages"].values(), *shard["packages.conda"].values()):
+        # to go faster, don't use PackageRecord, record.combined_depends, or
+        # MatchSpec
+        record = PackageRecord(**maybe_unpack_record(package))
+        mentioned.add(record.name)
+        mentioned.update(spec.name for spec in record.combined_depends)
+    return mentioned
+
+
+class ShardsIndex(TypedDict):
+    info: RepodataInfo
+    repodata_version: int
+    removed: list[str]
+    shards: dict[str, bytes]
+
+
+class ShardLike:
+    """
+    Present a "classic" repodata.json as per-package shards.
+    """
+
+    def __init__(self, repodata: dict, url: str = ""):
+        """
+        url: for debugging; not used by this class.
+        """
+        all_packages = {
+            "packages": repodata.pop("packages", {}),
+            "packages.conda": repodata.pop("packages.conda", {}),
+        }
+        self.repodata = repodata  # without packages, packages.conda
```

Comment:
Maybe rename to "package_less_repodata"?

Reply:
https://github.com/conda/conda-libmamba-solver/pull/686/files#diff-09617837bcdab07fd05c2faebb3585caab01a13ee029e1af4be262936ba0b4b7R116 "repodata_no_packages"